graphql-kotlin-schema-generator/src/main/kotlin/com/expediagroup/graphql/generator/state/TypesCache.kt

```diff
@@ -90,17 +90,23 @@ internal class TypesCache(private val supportedPackages: List<String>) {
 
     private fun isTypeNotSupported(type: KType): Boolean = supportedPackages.none { type.qualifiedName.startsWith(it) }
 
-    fun buildIfNotUnderConstruction(kClass: KClass<*>, build: (KClass<*>) -> GraphQLType): GraphQLType {
-        val cachedType = cache[kClass.getSimpleName()]
+    internal fun buildIfNotUnderConstruction(kClass: KClass<*>, inputType: Boolean, build: (KClass<*>) -> GraphQLType): GraphQLType {
+        if (kClass.isListType()) {
+            return build(kClass)
+        }
+
+        val cacheKey = TypesCacheKey(kClass.starProjectedType, inputType)
+        val cachedType = get(cacheKey)
         return when {
-            cachedType != null -> cachedType.graphQLType
-            typeUnderConstruction.contains(kClass) -> GraphQLTypeReference.typeRef(kClass.getSimpleName())
+            cachedType != null -> cachedType
+            typesUnderConstruction.contains(cacheKey) -> GraphQLTypeReference.typeRef(kClass.getSimpleName(inputType))
             else -> {
-                typeUnderConstruction.add(kClass)
+                typesUnderConstruction.add(cacheKey)
                 val newType = build(kClass)
-                val key = TypesCacheKey(kClass.createType(), false)
-                put(key, KGraphQLType(kClass, newType))
-                typeUnderConstruction.remove(kClass)
+                if (newType !is GraphQLTypeReference) {
```

Comment:
guess this is the fix for the #527 ?

Reply:
This one line doesn't solve the issue. It was also causing bugs to appear in the federation code which is why I needed to refactor the schema generator to be able to work without having to directly call `interfaceType` and `objectType`